### PR TITLE
CNV-8112: Added info about PCI host devices 

### PR DESCRIPTION
**Pull request by sjhala-ccs**
This PR addresses [CNV-8112](https://issues.redhat.com/browse/CNV-8112)

This content builds on top of what was merged in #32247 for CNV-11006. Modified the "Configuring PCI passthrough" assembly to add information about exposing PCI host devices in the cluster and the workflow to remove PCI devices from the cluster.

Applies to 4.8

**Preview builds**: 

- [Configuring PCI passthrough](https://deploy-preview-33108--osdocs.netlify.app/openshift-enterprise/latest/virt/virtual_machines/advanced_vm_management/virt-configuring-pci-passthrough.html)
- [Exposing PCI host devices in the cluster using the CLI](https://deploy-preview-33108--osdocs.netlify.app/openshift-enterprise/latest/virt/virtual_machines/advanced_vm_management/virt-configuring-pci-passthrough?utm_source=github&utm_campaign=bot_dp#virt-adding-pci-device-hyperconverged-operator_virt-configuring-pci-passthrough)
- [Removing PCI host devices from the cluster using the CLI](https://deploy-preview-33108--osdocs.netlify.app/openshift-enterprise/latest/virt/virtual_machines/advanced_vm_management/virt-configuring-pci-passthrough?utm_source=github&utm_campaign=bot_dp#virt-removing-pci-device-from-cluster_virt-configuring-pci-passthrough)

Ready for SME review: @fabiand @nunnatsa @vladikr 
Requested QE review from Issac Besso (tagged in Jira)

### CODE REVIEW

File: modules/virt-about-pci-passthrough.adoc [hunk 1]
```diff
@@ -3,8 +3,53 @@
...
+----
+<1> The host devices that are permitted to be used in the cluster.
+<2> The list of PCI devices available on the node.
+<3> The `vendor-ID` and the `device-ID` required to identify the PCI device.
+<4> The name of the first PCI host device that is enabled by default.
+<5> The name of the second PCI host device that is enabled by default.
+
+[NOTE]
+====
+You can add custom PCI host devices to be used in the cluster; however, the two default host devices cannot be removed. Disable the default devices by setting the `disabled` flag to `true`.
```
Comment (nunnatsa) [inline modules/virt-about-pci-passthrough.adoc:35]: I would add "if needed" or something like that. I mean - we think the defaults are good and useful - that why we put them there. so disable them only if needed: 
> however, the two default host devices cannot be removed. If needed, disable the default devices by setting the `disabled` flag to `true`, instead of removing them.

File: modules/virt-adding-pci-device-hyperconverged-operator.adoc [hunk 2]
```diff
@@ -25,23 +25,34 @@ metadata:
 spec:
   permittedHostDevices: <1>
     pciHostDevices: <2>
-      - pciVendorSelector: 10de:1eb8 <3>
-        resourceName: "nvidia.com/TU104GL_Tesla_T4" <4>
+    - pciDeviceSelector: "10DE:1DB6"  <3>
+      resourceName: "nvidia.com/GV100GL_Tesla_V100" <4>
+    - pciDeviceSelector: "10DE:1EB8"
+      resourceName: "nvidia.com/TU104GL_Tesla_T4"
+      disabled: true <5>
```
Comment (nunnatsa) [inline modules/virt-adding-pci-device-hyperconverged-operator.adoc:32]: not sure we need the `disable` flag in this example. Not sure we don't either...

**Review by nunnatsa — CHANGES_REQUESTED** [review]
This section only refers to PCI host devices. What about the mediated devices? Please see here for details; https://github.com/kubevirt/hyperconverged-cluster-operator/blob/main/docs/cluster-configuration.md#listing-permitted-host-devices

File: modules/virt-about-pci-passthrough.adoc (hunk 1 above)
Comment (sjhala-ccs) [reply modules/virt-about-pci-passthrough.adoc:35]: Made the change

File: modules/virt-adding-pci-device-hyperconverged-operator.adoc (hunk 2 above)
Comment (sjhala-ccs) [reply modules/virt-adding-pci-device-hyperconverged-operator.adoc:32]: Removed the `disabled` flag from this example

**Comment by sjhala-ccs** [comment]
> This is not output. This is an editing session.
> 
> Maybe we want to change line 14 - 42 by a `oc patch` command? will it be clearer? then we can add an `oc get HyperConverged` command for verification?

I can remove the term `output` from line 17 if that makes more sense. The way we use `Example output` is that the code block below is the output of the `oc` command that you run in the previous code block, in this case the `oc edit` command.

**Comment by sjhala-ccs** [comment]
@nunnatsa PTAL https://deploy-preview-33108--osdocs.netlify.app/openshift-enterprise/latest/virt/virtual_machines/advanced_vm_management/virt-configuring-pci-passthrough?utm_source=github&utm_campaign=bot_dp#virt-adding-pci-device-hyperconverged-operator_virt-configuring-pci-passthrough

Are the procedure steps clearer? If not then please provide detailed syntax and expected output for the `oc patch` command.

**Comment by sjhala-ccs** [comment]
Removed the text about default PCI host devices and made minor edits to the procedure steps for clarity.

**Comment by nunnatsa** [comment]
Personally, I find the `oc edit` method easier to do, but harder to implement. So here is the same in the jsonpatch way.

json patch is a bit tricky. if you want to add some sub-element that is not there yet, you need to first add the sub-element. So there is a difference between the first device and the rest of them. Trying to use the first one just to add a second device, will replace the whole element, including the existing device.

Here is a command to add the first device:
```bash
oc patch -n openshift-cnv HyperConverged kubevirt-hyperconverged --type=json -p='[{"op":"add","path":"/spec", "value": {"permittedHostDevices": {"pciHostDevices": [{"pciDeviceSelector":"10DE:1EB8","resourceName":"nvidia.com/TU104GL_Tesla_T4"}]}}}]'
```

To add a device to the existing list, use this:
```bash
oc patch -n openshift-cnv HyperConverged kubevirt-hyperconverged --type=json -p='[{"op":"add","path":"/spec/permittedHostDevices/pciHostDevices/-", "value": {"pciDeviceSelector":"10DE:1DB6","resourceName":"nvidia.com/GV100GL_Tesla_V100"}}]'
```

To remove a device, use this syntax (in this example - removing the first device in the list:
```
oc patch -n openshift-cnv HyperConverged kubevirt-hyperconverged --type=json -p='[{"op":"remove","path":"/spec/permittedHostDevices/pciHostDevices/0"}]'
```

**Comment by sjhala-ccs** [comment]
@fabiand @vladikr @nunnatsa Can someone PTAL and approve these changes if they look good? Thank you.

File: modules/virt-about-pci-passthrough.adoc [hunk 3]
```diff
@@ -3,8 +3,35 @@
...
+metadata:
+  name: kubevirt-hyperconverged
+spec:
+  permittedHostDevices: <1>
+    pciHostDevices: <2>
+    - pciDeviceSelector: "10DE:1DB6" <3>
+      resourceName: "nvidia.com/GV100GL_Tesla_V100" <4>
+    - pciDeviceSelector: "10DE:1EB8"
+      resourceName: "nvidia.com/TU104GL_Tesla_T4"
+...
```
Comment (fabiand) [inline modules/virt-about-pci-passthrough.adoc:25]: Why do we show these changes in the about section? Isn't this redundant to the snippet in the "Exposing PCI host devices in the cluster using the CLI" section?

**Review by fabiand — COMMENTED** [review]
lgtm except the confusion about the snippet in the about section

File: modules/virt-about-pci-passthrough.adoc (hunk 3 above)
Comment (sjhala-ccs) [reply modules/virt-about-pci-passthrough.adoc:25]: @fabiand The reason I mentioned the two Nvidia devices upfront is because, from what I understand, the admin does not need to go through the procedures to add kernel arguments and bind the devices to the VFIO driver. They can directly edit the HCO CR and expose these devices in the cluster. For any other PCI host device, the admin has to go through the whole process. I wanted to point out that difference.

Also, the example snippet in the "Exposing PCI host devices..." section shows a user-added device with the `externalResourceProvider `field set to `true` which is not used for the Nvidia devices. 

Does that make sense?

Comment (vladikr) [reply modules/virt-about-pci-passthrough.adoc:25]: > @fabiand The reason I mentioned the two Nvidia devices upfront is because, from what I understand, the admin does not need to go through the procedures to add kernel arguments and bind the devices to the VFIO driver. They can directly edit the HCO CR and expose these devices in the cluster. For any other PCI host device, the admin has to go through the whole process. I wanted to point out that difference.
> 
@sjhala-ccs Unfortunately, that's incorrect. Any device PCI device that is intended to be assignable to Virtual Machines needs to be bound to the VFIO driver. This includes these two devices.

> Also, the example snippet in the "Exposing PCI host devices..." section shows a user-added device with the `externalResourceProvider `field set to `true` which is not used for the Nvidia devices.

The `externalResourceProvider ` field indicates whether this resource is being provided by an external device plugin. If it's set to `true`, KubeVirt/CNV will only permit the usage of this device in the cluster but will leave the allocation and monitoring to an external device plugin.



> 
> Does that make sense?

File: modules/virt-exposing-pci-device-in-cluster-cli.adoc [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+kind: HyperConverged
+metadata:
+  name: kubevirt-hyperconverged
+  namespace: openshift-cnv
+spec:
+  permittedHostDevices:
+    pciHostDevices:
+    - pciDeviceSelector: "8086:6F54" <1>
+      resourceName: "intel.com/qat" <2>
+      externalResourceProvider: true <3>
```
Comment (vladikr) [inline modules/virt-exposing-pci-device-in-cluster-cli.adoc:33]: Perhaps we shouldn't mention this here. In most cases, there is no need to use an external device plugin. 
Maybe we can just briefly mention that such an option exists.

`In case, the device requires a specialized device plugin, please add the externalResourceProvider: true  to the device configuration section` ?

**Comment by sjhala-ccs** [comment]
@fabiand @vladikr I have updated the PR based on your suggestions. Please take a look. Thank you.

**Comment by fabiand** [comment]
Let's please address the last comment from Vlaidk, then we are good

/lgtm

**Comment by sjhala-ccs** [comment]
> Let's please address the last comment from Vlaidk, then we are good

Regarding Vladik's comment, I have mentioned in the callout text that this is an optional field and described the functionality when it is set to true.

File: modules/virt-exposing-pci-device-in-cluster-cli.adoc (hunk 4 above)
Comment (fabiand) [reply modules/virt-exposing-pci-device-in-cluster-cli.adoc]: Vladik, yes it's rare, but we might want to keep it for completeness. Otherwise it will not pop-up anywhere in the docs.

Comment (vladikr) [reply modules/virt-exposing-pci-device-in-cluster-cli.adoc]: Agreed

**Comment by vladikr** [comment]
Everything looks good to me now
Thank you @sjhala-ccs
/lgtm

File: modules/virt-removing-pci-device-from-cluster-cli.adoc [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+To remove a PCI host device from the cluster, delete the information for that device from the `HyperConverged` custom resource (CR).
+
+
+.Procedure
+. Edit the `HyperConverged` CR in YAML format using the default editor by running the following command:
++
+[source,terminal]
+----
+$ oc edit hyperconverged kubevirt-hyperconverged -n openshift-cnv -o yaml
```
Comment (kbidarkar) [inline modules/virt-removing-pci-device-from-cluster-cli.adoc:16]: We should drop `-o yaml` here, this shouldn't be required here.

File: modules/virt-exposing-pci-device-in-cluster-cli.adoc [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+To enable the availability of PCI host devices in the cluster, add details about the PCI devices to the `HyperConverged` custom resource (CR). The PCI devices are exposed and permitted to be used in the cluster.
+
+
+.Procedure
+. Edit the `HyperConverged` CR in YAML format using the default editor by running the following command:
++
+[source,terminal]
+----
+$ oc edit hyperconverged kubevirt-hyperconverged -n openshift-cnv -o yaml
```
Comment (kbidarkar) [inline modules/virt-exposing-pci-device-in-cluster-cli.adoc:16]: We should drop `-o yaml` here, this shouldn't be required here.

**Review by kbidarkar — CHANGES_REQUESTED** [review]

**Comment by kbidarkar** [comment]
Apart from dropping of `-o yaml` , LGTM

File: modules/virt-removing-pci-device-from-cluster-cli.adoc (hunk 5 above)
Comment (sjhala-ccs) [reply modules/virt-removing-pci-device-from-cluster-cli.adoc:16]: @kbidarkar Is the object configuration in YAML format by default? Is that why we don't need to specify `-o yaml`?

File: modules/virt-about-pci-passthrough.adoc [hunk 7]
```diff
@@ -3,8 +3,8 @@
 // * virt/virtual_machines/advanced_vm_management/virt-configuring-pci-passthrough.adoc
 
 [id="virt-about_pci-passthrough_{context}"]
-= About preparing nodes and the cluster for PCI passthrough
+= About preparing a host device for PCI passthrough
 
-The Peripheral Component Interconnect (PCI) passthrough feature enables you to access and manage hardware devices from a virtual machine. When PCI passthrough is configured, the PCI devices function as if they were physically attached to the virtual machine.
+To prepare a host device for PCI passthrough using the CLI, create a `MachineConfig` object and add kernel arguments to enable the Input-Output Memory Management Unit (IOMMU). Bind the PCI device to the Virtual Function I/O (VFIO) driver and then expose it in the cluster by editing the `permittedHostDevices` field of the `HyperConverged` custom resource (CR). The `permittedHostDevices` list is empty when you first install the HyperConverged Cluster Operator.
```
Comment (ousleyp) [inline modules/virt-about-pci-passthrough.adoc:8]: This is very much a suggestion, but maybe:

s/HyperConverged Cluster Operator/`{VirtProductName}` Operator

The install docs tend to avoid the former: https://docs.openshift.com/container-platform/4.8/virt/install/installing-virt-web.html

File: modules/virt-about-pci-passthrough.adoc [hunk 8]
```diff
@@ -3,8 +3,8 @@
...
 
-To configure a hardware device as a host device for PCI passthrough, create a `MachineConfig` object file and add kernel arguments to enable Input-Output Memory Management Unit (IOMMU). Bind the PCI device to the Virtual Function I/O (VFIO) driver by creating a `MachineConfig` object file. Then, add the PCI device to the HyperConverged Operator.
+To remove a PCI host device from the cluster using the CLI, delete the PCI device information from the `HyperConverged` CR.
```
Comment (ousleyp) [inline modules/virt-about-pci-passthrough.adoc:10]: s/using/by using

File: modules/virt-about-pci-passthrough.adoc (hunk 7 above)
Comment (ousleyp) [inline modules/virt-about-pci-passthrough.adoc:8]: s/using the CLI/by using the CLI

File: modules/virt-exposing-pci-device-in-cluster-cli.adoc [hunk 9]
```diff
@@ -0,0 +1,91 @@
+// Module included in the following assemblies:
+//
+// * virt/virtual_machines/advanced_vm_management/virt-configuring-pci-passthrough.adoc
+
+[id="virt-exposing-pci-device-in-cluster-cli_{context}"]
+= Exposing PCI host devices in the cluster using the CLI
+
+To enable the availability of PCI host devices in the cluster, add details about the PCI devices to the `HyperConverged` custom resource (CR). The PCI devices are exposed and permitted to be used in the cluster.
```
Comment (ousleyp) [inline modules/virt-exposing-pci-device-in-cluster-cli.adoc:8]: "To enable the availability of" strikes me as a bit wordy, and this paragraph seems like it could be shortened to just one sentence. I wonder if something like this could work (completely optional):

"To expose PCI devices to the cluster so that they can be used, add them to the `spec.permittedHostDevices` stanza of the `HyperConverged` custom resource (CR)."

File: modules/virt-exposing-pci-device-in-cluster-cli.adoc [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+
+.Procedure
+. Edit the `HyperConverged` CR in YAML format using the default editor by running the following command:
```
Comment (ousleyp) [inline modules/virt-exposing-pci-device-in-cluster-cli.adoc:12]: I might s/in YAML format using the default editor/in your default editor

(mostly because we don't usually specify that it's going to be a YAML)

File: modules/virt-exposing-pci-device-in-cluster-cli.adoc [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    - pciDeviceSelector: "8086:6F54"
+      resourceName: "intel.com/qat"
+      externalResourceProvider: true <5>
+...
+----
+<1> The host devices that are permitted to be used in the cluster.
+<2> The list of PCI devices available on the node.
+<3> The `vendor-ID` and the `device-ID` required to identify the PCI device.
+<4> The name of a PCI host device.
+<5> Optional field. When set to `true` indicates that the resource is provided by an external device plug-in. {VirtProductName} allows the usage of this device in the cluster but leaves the allocation and monitoring to an external device plug-in.
```
Comment (ousleyp) [inline modules/virt-exposing-pci-device-in-cluster-cli.adoc:45]: s/Optional field. When set to `true` indicates/Optional: Setting this field to `true` indicates

File: modules/virt-exposing-pci-device-in-cluster-cli.adoc [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+====
+
+. Save your changes and exit the editor.
+
+.Verification
+* Verify that the PCI host devices were added to the node by running the following command. The example output shows that there is one device each associated with the `nvidia.com/GV100GL_Tesla_V100`, `nvidia.com/TU104GL_Tesla_T4`, and `intel.com/qat` resource names.
++
+[source,terminal]
+----
+$ oc describe node <node1>
```
Comment (ousleyp) [inline modules/virt-exposing-pci-device-in-cluster-cli.adoc:59]: I think that instead of `<node1>` this should be something like `<node_name>`. If you want to use `node1` as an example name I would take the `<>` out.

File: modules/virt-removing-pci-device-from-cluster-cli.adoc [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+// * virt/virtual_machines/advanced_vm_management/virt-configuring-pci-passthrough.adoc
+
+[id="virt-removing-pci-device-from-cluster_{context}"]
+= Removing PCI host devices from the cluster using the CLI
+
+To remove a PCI host device from the cluster, delete the information for that device from the `HyperConverged` custom resource (CR).
+
+
+.Procedure
+. Edit the `HyperConverged` CR in YAML format using the default editor by running the following command:
```
Comment (ousleyp) [inline modules/virt-removing-pci-device-from-cluster-cli.adoc:12]: Same comment as in the other module

**Review by ousleyp — COMMENTED** [review]
Overall, I think this looks great! Just a few comments/suggestions.

File: modules/virt-removing-pci-device-from-cluster-cli.adoc (hunk 5 above)
Comment (sjhala-ccs) [reply modules/virt-removing-pci-device-from-cluster-cli.adoc:16]: Removed `-o yaml` from the command
